README.md
# HaskellVideoGame
A simple video game implemented in Haskell

I wrote this game after 4 days of learning Haskell, in a week, for a class.

Run using `cabal run`


ChangeLog.md
# Revision history for HaskellVideoGame

## 0.1.0.0 -- YYYY-mm-dd

* First version. Released on an unsuspecting world.


Game.hs
module Game where

import Entity

import Graphics.Gloss
-- import Graphics.Gloss.Interface.Pure.Game
getDecimal :: RealFrac a => a -> a
getDecimal = snd . properFraction

-- | Data describing the state of the game
data GameState = Game
  { player :: GameObject -- ^ The player
  , enemies :: [GameObject] -- ^ A list of the enemies
  , bullets :: [Bullet] -- ^ A list of bullets on screen
  , timeElapsed :: Float -- ^ The total elapsed time of the game
  , textShown :: Maybe String -- ^ Text displayed on screen
  , textScale :: Float
  , isPaused :: Bool
  , playerAlive :: Bool
  , gameStep :: GameState -> GameState
  , currentLevel :: GameState -> GameState
  }

initialState :: GameState
initialState = Game
  { player = initialPlayer
  , enemies = []
  , bullets = []
  , timeElapsed = 0
  , textShown = Nothing
  , textScale = 0
  , isPaused = False
  , playerAlive = True
  , gameStep = tutorial_0
  , currentLevel = tutorial_0
  }
----------------------------------------------------------------------------------------
-- | collision order matters -- collide_enemies will remove enemies and collide_bullets will remove bullets
update :: Float -> GameState -> GameState
update seconds = collide_bullets . collide_enemies . collide_player . run_game seconds
----------------------------------------------------------------------------------------
-- | GameState modifying functions
distance :: Point -> Point -> Float
distance (x1, y1) (x2, y2) = sqrt (x'*x' + y'*y')
  where
    x' = x2 - x1
    y' = y2 - y1
-- | function to check if any of our hero's bullets have hit an enemy 
collide_enemies :: GameState -> GameState
collide_enemies game = if playerAlive game then game { enemies = enemies'' } else game
  where
     -- If an enemy is too close to a bullet, apply damage (same order as bullets')
    enemies' = 
      foldr 
        (\enemy -> (\r ->
          [enemy { curHealth = (curHealth enemy) 
            - (sum 
                [damage b | b <- (filter (\bullet -> (distance (position enemy) (bullet_pos bullet)) < (hit_radius enemy) + (bullet_hit_radius bullet))
                  (bullets game))]) }] ++ r )) [] (enemies game)
    enemies'' =
      foldr 
        (\enemy -> (\r ->
          if (distance (position enemy) (position (player game))) < (hit_radius enemy) + (hit_radius (player game))
          then r else [enemy] ++ r )) [] enemies'
    -- If an enemy hits a player, remove it
-- | function to check if any bullets need to be cleaned up after hitting an enemy
collide_bullets :: GameState -> GameState
collide_bullets game = game { bullets = bullets' }
  where
    -- If a bullet hit an enemy, remove it (Order n*m where n is num bullets and m is num enemies)
    bullets' = 
      foldr (\bullet -> (\r ->
              if any 
                (\enemy -> (distance (position enemy) (bullet_pos bullet)) < (hit_radius enemy) + (bullet_hit_radius bullet)) 
                (enemies game)
              then r else [bullet] ++ r)) [] (bullets game)
-- | function for colliding the hero
collide_player :: GameState -> GameState
collide_player game = if playerAlive game then game { player = player' } else game
  where -- | if there is an enemy that hits our player, lose health and move to origin
    prev = curHealth (player game)
    hitEnemy = length (filter (\enemy -> 
                                (distance (position enemy) (position (player game)) < 
                                ((hit_radius enemy) + (hit_radius (player game)))))
                              (enemies game)) > 0
    curHealth' = if hitEnemy then prev - 1 else prev
    position' = if hitEnemy then (0, 0) else position (player game)
    player' = (player game) { curHealth = curHealth', position = position' } 

----------------------------------------------------------------------------------------
-- Game running functions
----------------------------------------------------------------------------------------
-- | Basic "run the game" function
run_game :: Float -> GameState -> GameState -- This ghastly statement is because movestep might change gamestep. Everything inside is linear time
run_game seconds game = (gameStep (move_step seconds game)) (move_step seconds game)
---------------------------------------------------------------------------------------
-- | Basic gameplay move everything function
move_step :: Float -> GameState -> GameState
move_step seconds game =
  if (isPaused game) then game else
  game { player = player'
       , enemies = enemies'
       , bullets = bullets'
       , timeElapsed = timeElapsed'
       , playerAlive = playerAlive'
       , gameStep = gameStep'
       }
    where  
      [player'] = if (playerAlive game) then move_player seconds (player game) else [(player game)]
      enemies' = foldr (\e -> (\r -> (step e) e ++ r)) [] (foldr (\e -> (\r -> move_enemy seconds e ++ r)) [] (enemies game))
      bullets' = foldr (\b -> (\r -> (bullet_step b) b ++ r)) [] 
                  (foldr (\b -> (\r -> move_bullets seconds b ++ r)) [] 
                    (bullets game))
      timeElapsed' = timeElapsed game + seconds
      playerAlive' = if (curHealth (player game)) <= 0 then False else True
      gameStep' = if playerAlive' then (gameStep game) else game_over (Just "Press r to retry")

spawn_enemies :: [SpawnInfo] -> (GameState -> GameState) -> (GameState -> GameState)
spawn_enemies si func
  | si == [] = \game -> game { gameStep = func }
  | True = \game ->
    game { enemies = (enemies game) ++ if st == Zoomer 
                                       then [generateZoomer loc f]
                                       else if st == Seeker
                                       then [generateSeeker loc f]
                                       else if st == Carrier
                                       then [generateCarrier loc f] 
                                       else []
         , gameStep = spawn_enemies (tail si) func }
      where 
        (SI loc f st) = head si
-- | Runs the game as normal for a given amount of time before the next function
-- Shamelessly stolen from instructor's example
run_for :: Float -> (GameState -> GameState) -> (GameState -> GameState)
run_for ticks f2 = \g ->
  if ticks == 0 then
    g { gameStep = f2 }
  else
    g { gameStep = run_for (ticks - 1) f2 }

show_text :: String -> (GameState -> GameState) -> (GameState -> GameState)
show_text text f = \g ->
  g { textShown = Just text , textScale = 1.0, gameStep = f }

-- | A function that displays a countdown to the game start | Assumes the game is running perfectly at 60 fps
countdown :: Int -> (GameState -> GameState) -> (GameState -> GameState)
countdown ticks f = \g ->
  if ticks == 0 then
    g { gameStep = f , textShown = Nothing }
  else 
    g { gameStep = countdown (ticks - 1) f , textShown = textShown' , textScale = 1 - (getDecimal (timeElapsed g)) }
  where 
    textShown' = Just (show (1 + quot ticks 60))

wait_clear :: (GameState -> GameState) -> (GameState -> GameState)
wait_clear f = \g ->
  if length (enemies g) == 0 then g { gameStep = f } else g 

reset_player :: (GameState -> GameState) -> (GameState -> GameState)
reset_player f = \g ->
  g { player = (player g) { curHealth = (maxHealth (player g)) , position = (0, 0) } 
    , gameStep = f 
    , playerAlive = True}

levelup :: (GameState -> GameState) -> (GameState -> GameState)
levelup level_x = \g ->
  g { gameStep = level_x , currentLevel = level_x }

game_over :: Maybe String -> (GameState -> GameState)
game_over text = \g ->
  g { playerAlive = False, textShown = text, textScale = 1.0 }

tutorial_0 :: GameState -> GameState
tutorial_0 = reset_player
           $ show_text "WASD to move"
           $ run_for 180
           $ show_text "Click to shoot"
           $ run_for 180
           $ show_text "Kill the zoomer"
           $ spawn_enemies tutorial_0_spawn
           $ wait_clear
           $ show_text "Good"
           $ reset_player
           $ run_for 90
           $ levelup tutorial_1

tutorial_1 :: GameState -> GameState
tutorial_1 = show_text "Press 2"
           $ run_for 180
           $ show_text "Starburst"
           $ run_for 90
           $ countdown 180
           $ spawn_enemies tutorial_1_spawn
           $ wait_clear
           $ show_text "Good"
           $ reset_player
           $ run_for 90
           $ levelup tutorial_2

tutorial_2 :: GameState -> GameState
tutorial_2 = show_text "Press 3"
           $ run_for 90
           $ show_text "Boomerang"
           $ run_for 90
           $ countdown 180
           $ spawn_enemies tutorial_1_spawn
           $ wait_clear
           $ show_text "Hit it backwards"
           $ run_for 90
           $ countdown 180
           $ spawn_enemies tutorial_1_spawn
           $ wait_clear
           $ show_text "Good"
           $ reset_player
           $ run_for 90
           $ levelup level_0

level_0 :: GameState -> GameState
level_0 = show_text "Good Luck"
        $ run_for 60
        $ countdown 180
        $ spawn_enemies level_0_spawn
        $ wait_clear
        $ reset_player
        $ show_text "Good"
        $ run_for 60
        $ levelup level_1

level_1 :: GameState -> GameState
level_1 = show_text "Dodge This"
        $ run_for 60
        $ countdown 180
        $ spawn_enemies level_1_spawn
        $ wait_clear
        $ show_text "Good"
        $ reset_player
        $ run_for 90
        $ levelup level_2

level_2 :: GameState -> GameState
level_2 = show_text "Again"
        $ run_for 60
        $ countdown 180
        $ spawn_enemies level_2_spawn
        $ wait_clear
        $ show_text "Hey nice"
        $ run_for 120
        $ show_text "That's all (ESC)"
        $ (\g -> g)


Render.hs
module Render where

import Game
import Entity

import Data.Maybe
import Graphics.Gloss
import Graphics.Gloss.Geometry.Angle

window :: Display
window = FullScreen

background :: Color
background = black

fps :: Int
fps = 60

-- | Derpy helper function for drawGameObject
hpRatio :: GameObject -> Float
hpRatio go = (curHealth go) / (maxHealth go)
showText :: Maybe String -> Float -> [Picture]
showText textToShow textScale
  | (isNothing textToShow) = []
  | True = if (isJust textToShow) 
            then ([ scale textScale textScale 
                 $ translate (-center) 0
                 $ color white 
                 $ text (fromJust textToShow)])
            else []
              where 
                center = (fromIntegral (length (fromJust textToShow))) * (40 * textScale)

-- | Draws game objects
drawGameObject :: GameObject -> Picture
drawGameObject go = 
  pictures [(uncurry translate (position go)
           $ rotate (facing go)
           $ color (entityColor go)
           $ lineLoop (model go)) 
           ,(uncurry translate (position go)
           $ color hpColor 
           $ polygon [(-15, -20), (-15, -22), (hpRemain, -22), (hpRemain, -20)])]
    where
      hpColor = 
        if (hpRatio go > 2.1/3.0) then green
        else if (hpRatio go > 1.1/3.0) then yellow
        else red
      hpRemain = (-15.0) + (30.0 * (hpRatio go))

-- | Draws Bullets
drawBullet :: Bullet -> Picture
drawBullet b = uncurry translate (bullet_pos b)
             $ rotate (bullet_fac b)
             $ (bullet_pic b)

drawBorders :: Picture
drawBorders = color red (rectangleWire 1500 1000)

-- | Function to turn the gamestate into a picture
render :: GameState -> Picture
render game =
  pictures ([playerPicture] ++ enemyPictures ++ bulletPictures ++ textPicture ++ [borders])
  where
    playerPicture = drawGameObject (player game)
    enemyPictures = map drawGameObject (enemies game)
    bulletPictures = map drawBullet (bullets game)
    textPicture = showText (textShown game) (textScale game)
    borders = drawBorders

Keyboard.hs
module Keyboard where

import Game
import Entity

import Graphics.Gloss.Interface.Pure.Game
import Graphics.Gloss.Geometry.Angle

-- | Enum I made to help keep track of wasd inputs
data MoveDir = UP | DOWN | LEFT | RIGHT | NOUP | NODOWN | NOLEFT | NORIGHT
  deriving (Enum, Eq)

----------------------------------------------------------------------------------------
-- Response functions
----------------------------------------------------------------------------------------
setPlayerVel :: MoveDir -> GameState -> GameState
setPlayerVel direction game 
  | direction == UP      = game { player = (player game) { move_direction = (x, y + 1) } }
  | direction == DOWN    = game { player = (player game) { move_direction = (x, y - 1) } }
  | direction == LEFT    = game { player = (player game) { move_direction = (x - 1, y) } }
  | direction == RIGHT   = game { player = (player game) { move_direction = (x + 1, y) } }
  | direction == NOUP    = game { player = (player game) { move_direction = (x, y - 1) } }
  | direction == NODOWN  = game { player = (player game) { move_direction = (x, y + 1) } }
  | direction == NOLEFT  = game { player = (player game) { move_direction = (x + 1, y) } }
  | direction == NORIGHT = game { player = (player game) { move_direction = (x - 1, y) } }
  where 
    (x, y) = move_direction (player game)

setPlayerWeapon :: Weapon -> GameState -> GameState
setPlayerWeapon weap game
  | weap == BASIC     = game { player = (player game) { weapon_type = BASIC , shoot_cooldown = 0 } }
  | weap == STARBURST = game { player = (player game) { weapon_type = STARBURST , shoot_cooldown = 0.5 }} -- half the cooldown so people can't just spam 2 to spam shoot
  | weap == BOOMERANG = game { player = (player game) { weapon_type = BOOMERANG , shoot_cooldown = 0.25 }}

playerTurnAndShoot :: (Float, Float) -> GameState -> GameState
playerTurnAndShoot (mouse_x, mouse_y) game = 
  game { player = (player game) { facing = degrees , shoot_cooldown = shoot_cooldown'} 
       , bullets = bullets' 
       }
  where
    p = player game
    (player_x, player_y) = position p
    dx = mouse_x - player_x 
    dy = mouse_y - player_y
    degrees = if dy > 0 then radToDeg( atan (dx/dy) ) else 180 + radToDeg (atan (dx/dy))
    bullets' = if (shoot_cooldown p) == 0 then 
                (bullets game) ++ 
                       if (weapon_type p) == BASIC     then generateBasicBullet degrees (player_x, player_y)
                  else if (weapon_type p) == STARBURST then generateStarburst   degrees (player_x, player_y) 
                  else if (weapon_type p) == BOOMERANG then generateBoomerang   degrees (player_x, player_y)
                  else []
               else (bullets game)
    shoot_cooldown' = if (shoot_cooldown p) == 0 then
                        if (weapon_type p) == BASIC then 0 else
                        if (weapon_type p) == STARBURST then 1 else
                        if (weapon_type p) == BOOMERANG then 0.5 else 0
                      else (shoot_cooldown p)

-- | Key listener function
handleKeys :: Event -> GameState -> GameState
handleKeys (EventKey key keyState _ mouse) game
-- Player movement
  | key == Char 'w' && keyState == Down = setPlayerVel UP      game
  | key == Char 'w' && keyState == Up   = setPlayerVel NOUP    game
  | key == Char 'a' && keyState == Down = setPlayerVel LEFT    game
  | key == Char 'a' && keyState == Up   = setPlayerVel NOLEFT  game
  | key == Char 's' && keyState == Down = setPlayerVel DOWN    game
  | key == Char 's' && keyState == Up   = setPlayerVel NODOWN  game
  | key == Char 'd' && keyState == Down = setPlayerVel RIGHT   game
  | key == Char 'd' && keyState == Up   = setPlayerVel NORIGHT game
-- Mouse click reacting
  | key == MouseButton LeftButton && keyState == Down && playerAlive game = playerTurnAndShoot mouse game
-- Pause
  | key == Char 'p' && keyState == Down && playerAlive game = game { isPaused = not (isPaused game) }
-- New Game
  | key == Char 'n' && keyState == Down = initialState
-- Retry
  | key == Char 'r' && keyState == Down = game { player = (player game) { curHealth = (maxHealth (player game))
                                                                        , position = (0, 0) }
                                               , gameStep = (currentLevel game)
                                               , enemies = []
                                               , textShown = Nothing }
-- Change weapons 1 - basic, 2 - starburst, 3 - TBD
  | key == Char '1' && keyState == Down = setPlayerWeapon BASIC     game
  | key == Char '2' && keyState == Down = setPlayerWeapon STARBURST game
  | key == Char '3' && keyState == Down = setPlayerWeapon BOOMERANG game

handleKeys _ game = game

Entity.hs
module Entity where

import Graphics.Gloss
import Graphics.Gloss.Interface.Pure.Game
import Graphics.Gloss.Geometry.Angle

-- | Data type for spawning large groups of enemies
data SpawnInfo = SI Point Float EnemyType
  deriving (Eq)
type EnemyGroup = [SpawnInfo]

data Weapon = BASIC | STARBURST | BOOMERANG | ENEMY
  deriving (Enum, Eq)

data EnemyType = Zoomer | Seeker | Carrier
  deriving (Enum, Eq)

-- | Bullets. They hit enemies.
data Bullet = Bullet 
  { lifespan :: Float -- ^ How long it fly
  , damage :: Float   -- ^ How hard it hit
  , bullet_pos :: Point -- ^ Where it be
  , bullet_vel :: Float -- ^ How fast it go
  , bullet_rot :: Float -- ^ How much it spin
  , bullet_dir :: Float -- ^ direction in degrees because I can't make up my mind
  , bullet_fac :: Float -- ^ Facing of the bullet for the purposes of drawing
  , bullet_step :: (Bullet -> [Bullet]) -- ^ The bullet behavior function
  , bullet_pic :: Picture -- ^ What the bullet looks like (Fancy composed pictures because we're special)
  , bullet_hit_radius :: Float -- ^ Because everything hits stuff 
  }
-- | Entities in the game - enemies, player, etc
data GameObject = GO
  { maxHealth :: Float     -- ^ The max health of the entity 
  , curHealth :: Float     -- ^ the current health of the entity
  , position :: Point -- ^ The position of the entity
  , velocity :: Float -- ^ How fast it goes
  , step :: (GameObject -> [GameObject]) -- ^ The update function of the object
  , shoot :: (Float -> Point -> [Bullet]) -- ^ The shoot function of the object 
  , shoot_cooldown :: Float -- ^ The cooldown between shots, in seconds.
  , weapon_type :: Weapon -- ^ The weapon being shot. Directly affects shoot_cooldown
  , model :: Path -- ^ The outline of the entity (using lineLoop)
  , entityColor :: Color -- ^ The color of the entity
  , facing :: Float -- ^ The direction of the front of the entity in degrees because the people that made gloss are heathens
  , move_direction :: Point -- ^ The direction of the motion as a vector
  , hit_radius :: Float -- ^ The radius of the hitbox (circle) of this game object
  }

-- | The player's ship. Pointy triangle boi
playerModel :: Path
playerModel = [(0, 15), (-7.5, -15), (0, -7.5), (7.5, -15)]
-- | A zoomer. Looks like a derpy plane
zoomer :: Path
zoomer = [(0, 20), (-5, 10), (-5, 0), (-30, -15), (-5, -15), (-5, -20), 
          (5, -20), (5, -15), (30, -15), (5, 0), (5, 10)]
-- | A seeker. Looks like a mean zoomer
seeker :: Path
seeker = [(0, 20), (-15, 10), (-5, 10), (-5, 0), (-30, -15), (-5, -15), (-5, -20), 
          (5, -20), (5, -15), (30, -15), (5, 0), (5, 10), (15, 10)]
-- | A carrier. Looks cool but its healthbar is derpy
carrier :: Path
carrier = [(0, 32), (-24, 20), (-4, 20), (-4, -24), (-16, -24), (-20, -32), 
           (20, -32), (16, -24), (4, -24), (4, 20), (24, 20)]
-- | Carrier
-- | A bullet. Looks like a rectangle
basicBullet :: Path
basicBullet = [(-1, -4), (-1, 4), (1, 4), (1, -4)]

starburst_helper :: Float -> Picture
starburst_helper b = rotate (b * 45)
             $ translate 0 6
             $ color (light yellow) (polygon basicBullet)

starburst_pic :: Picture
starburst_pic = pictures (basics ++ [body])
  where
    basics = [ starburst_helper 0
             , starburst_helper 1
             , starburst_helper 2
             , starburst_helper 3
             , starburst_helper 4
             , starburst_helper 5
             , starburst_helper 6
             , starburst_helper 7
             ]
    body = color (dark red) (circleSolid 4)

boomerang_pic :: Picture
boomerang_pic = pictures [l, r]
  where
    l = color (light yellow) (polygon [(0, 3), (-1, 2), (4, -3), (5, -2)])
    r = color (light yellow) (polygon [(0, 3), (-5, -2), (-4, -3), (1, 2)])

tutorial_0_spawn :: EnemyGroup
tutorial_0_spawn = [SI (0, 500) 180 Zoomer]

tutorial_1_spawn :: EnemyGroup
tutorial_1_spawn = [ SI (0, 500) 135 Zoomer
                   , SI (-500, 0) 135 Zoomer
                   , SI (0, -500) 135 Zoomer
                   , SI (500, 0) 135 Zoomer
                   ]

level_0_spawn :: EnemyGroup
level_0_spawn = [SI (-250, 250)  90  Zoomer
                ,SI (250, 250)   180 Zoomer
                ,SI (250, -250)  270 Zoomer
                ,SI (-250, -250) 0   Zoomer
                ,SI (-250, 0)    45  Zoomer
                ,SI (250, 0)     225 Zoomer
                ,SI (0, -250)    315 Zoomer
                ,SI (0, 250)     135 Zoomer
                ]

level_1_spawn :: EnemyGroup
level_1_spawn = [SI (-250, 250)  90  Zoomer
                ,SI (250, 250)   180 Zoomer
                ,SI (250, -250)  270 Zoomer
                ,SI (-250, -250) 0   Zoomer
                ,SI (-250, 0)    45  Seeker
                ,SI (250, 0)     225 Seeker
                ,SI (0, -250)    315 Seeker
                ,SI (0, 250)     135 Seeker
                ]

level_2_spawn :: EnemyGroup
level_2_spawn = [SI (-250, 250)  90  Carrier
                ,SI (250, 250)   180 Carrier
                ,SI (-250, 0)    45  Seeker
                ,SI (250, 0)     225 Seeker
                ,SI (0, -250)    315 Seeker
                ,SI (0, 250)     135 Seeker
                ]

-- | The initial state of the player
initialPlayer :: GameObject
initialPlayer = GO
  { maxHealth = 3
  , curHealth = 3
  , position = (0, 0)
  , velocity = 500
  , step = (\a -> [a]) -- no effect, player does what they want (for now)
  , shoot = generateBasicBullet
  , shoot_cooldown = 0
  , weapon_type = BASIC
  , model = playerModel
  , entityColor = light blue
  , facing = 180
  , move_direction = (0, 0)
  , hit_radius = 10
  }
-- | A function that spits out a zoomer
generateZoomer :: Point -> Float -> GameObject
generateZoomer pos face = GO 
  { maxHealth = 2
  , curHealth = 2
  , position = pos
  , velocity = 525
  , step = (\a -> [a]) -- no effect
  , shoot = no_shoot
  , shoot_cooldown = 0
  , weapon_type = ENEMY
  , model = zoomer
  , entityColor = light red
  , facing = face
  , move_direction = (0, 0)
  , hit_radius = 20
  }

generateSeeker :: Point -> Float -> GameObject
generateSeeker pos face = GO
  { maxHealth = 3
  , curHealth = 3
  , position = pos
  , velocity = 0
  , step = seeker_step
  , shoot = no_shoot
  , shoot_cooldown = 0
  , weapon_type = ENEMY
  , model = seeker
  , entityColor = light magenta
  , facing = face
  , move_direction = (0, 0)
  , hit_radius = 20
  }

generateCarrier :: Point -> Float -> GameObject
generateCarrier pos face = GO
  { maxHealth = 10
  , curHealth = 10
  , position = pos
  , velocity = 450
  , step = carrier_step
  , shoot = no_shoot
  , shoot_cooldown = 0
  , weapon_type = ENEMY
  , model = carrier
  , entityColor = light orange
  , facing = face
  , move_direction = (0, 0)
  , hit_radius = 32
  }
----------------------------------------------------------------------------------------
-- Bullet building blocks
----------------------------------------------------------------------------------------
-- | Some enemies don't shoot
no_shoot :: Float -> Point -> [Bullet]
no_shoot _ _ = []
-- | Pew pew
generateBasicBullet :: Float -> Point -> [Bullet]
generateBasicBullet dir pos = 
  [Bullet 
  { lifespan = 10
  , damage = 1
  , bullet_pos = pos
  , bullet_vel = 800
  , bullet_dir = dir
  , bullet_rot = 0
  , bullet_fac = dir
  , bullet_step = (\b -> [b])
  , bullet_pic = color (light yellow) (polygon basicBullet)
  , bullet_hit_radius = 0 -- Trust me on this one it really doesn't matter
  }]
-- | Pew ... pewpewpwepwepwpewpewpewpepwepw
generateStarburst :: Float -> Point -> [Bullet]
generateStarburst dir pos = 
  [Bullet
  { lifespan = 10
  , damage = 2
  , bullet_pos = pos
  , bullet_vel = 600
  , bullet_dir = dir
  , bullet_rot = 6
  , bullet_fac = dir
  , bullet_step = startburst_step
  , bullet_pic = starburst_pic
  , bullet_hit_radius = 8
  }]
-- | Pew ... weeeeeeeeeeep
generateBoomerang :: Float -> Point -> [Bullet]
generateBoomerang dir pos =
  [Bullet
  { lifespan = 10
  , damage = 1
  , bullet_pos = pos
  , bullet_vel = 600
  , bullet_dir = dir
  , bullet_rot = 6
  , bullet_fac = dir
  , bullet_step = boomerang_step
  , bullet_pic = boomerang_pic
  , bullet_hit_radius = 4
  }]
----------------------------------------------------------------------------------------
-- Movement basic functions
----------------------------------------------------------------------------------------
-- | Bullet basic move function, NEVER modifies bullet_step. I made that mistake already with game over, might fix it later
move_bullets :: Float -> Bullet -> [Bullet]
move_bullets seconds b 
  | lifespan b - seconds <= 0 = [] -- remove the bullet if it's somehow still around after its lifespan
  | (x' > 750) || (x' < -750) || (y' > 500) || (y' < -500) = [] -- remove the bullet if it goes off screen
  | lifespan b - seconds > 0  = [b { lifespan = lifespan', bullet_pos = position' , bullet_fac = bullet_fac' }]
    where
      (x, y) = bullet_pos b
      v = bullet_vel b
      dir = bullet_dir b
      x' = x + v * sin(degToRad(dir)) * seconds
      y' = y + v * cos(degToRad(dir)) * seconds
      lifespan' = lifespan b - seconds
      position' = (x', y')
      bullet_fac' = (bullet_fac b) + (bullet_rot b)
-- | Player basic move function
move_player :: Float -> GameObject -> [GameObject]
move_player seconds go = [go { position = position' , shoot_cooldown = shoot_cooldown'} ]
  where
    v = velocity go
    (x, y) = position go
    (theta_x, theta_y) = move_direction go
    hyp = sqrt ((theta_x*theta_x)+(theta_y*theta_y)) -- for normalization of the move direction vector
    x' = if hyp > 0 then x + v * seconds * (theta_x / hyp) else x
    y' = if hyp > 0 then y + v * seconds * (theta_y / hyp) else y
    position' = if x' > 730    then (730, y')    else
                if x' < (-730) then ((-730), y') else
                if y' > 480    then (x', 480)    else
                if y' < (-480) then (x', (-480)) else (x', y')
    shoot_cooldown' = if (shoot_cooldown go) > 0 then max 0 ((shoot_cooldown go) - seconds) else 0
-- | zoomer basic move function, also removes the enemy if it's dead
move_enemy :: Float -> GameObject -> [GameObject]
move_enemy seconds go 
  | curHealth go <= 0 = []
  | True = [go { position = position' }]
      where
        v = velocity go
        (x, y) = position go
        dir = facing go
        x' = x + v * seconds * sin(degToRad(dir))
        y' = y + v * seconds * cos(degToRad(dir))
        position' = if x' > 730.0 then ((-730.0), y') else
                    if x' < (-730.0) then (730.0, y') else
                    if y' > 480.0 then (x', (-480.0)) else
                    if y' < (-480.0) then (x', 480.0) else (x', y')
-----------------------------------------------------------------------------------------
-- Bullet building blocks
-----------------------------------------------------------------------------------------
slowToZero :: (Bullet -> [Bullet]) -> (Bullet -> [Bullet])
slowToZero f = \b ->
  if bullet_vel b == 0 then [ b { bullet_step = f } ]
  else [ b { bullet_vel = max ((bullet_vel b) - 10) 0 } ]

spinUpTo :: Float -> (Bullet -> [Bullet]) -> (Bullet -> [Bullet])
spinUpTo total f = \b ->
  if (bullet_rot b) >= total then [ b { bullet_step = f } ]
  else [ b { bullet_rot = (bullet_rot b) + 0.25 } ]

burst :: (Bullet -> [Bullet]) -> (Bullet -> [Bullet])
burst f = \b -> generateBasicBullet (bullet_fac b + 0)     (bullet_pos b)
             ++ generateBasicBullet (bullet_fac b + 45)    (bullet_pos b)
             ++ generateBasicBullet (bullet_fac b + 90)    (bullet_pos b)
             ++ generateBasicBullet (bullet_fac b + 135)   (bullet_pos b)
             ++ generateBasicBullet (bullet_fac b + 180)   (bullet_pos b)
             ++ generateBasicBullet (bullet_fac b + 225)   (bullet_pos b)
             ++ generateBasicBullet (bullet_fac b + 270)   (bullet_pos b)
             ++ generateBasicBullet (bullet_fac b + 315)   (bullet_pos b)
             ++ generateBasicBullet (bullet_fac b + 22.5)  (bullet_pos b)
             ++ generateBasicBullet (bullet_fac b + 67.5)  (bullet_pos b)
             ++ generateBasicBullet (bullet_fac b + 112.5) (bullet_pos b)
             ++ generateBasicBullet (bullet_fac b + 157.5) (bullet_pos b)
             ++ generateBasicBullet (bullet_fac b + 200.5) (bullet_pos b)
             ++ generateBasicBullet (bullet_fac b + 247.5) (bullet_pos b)
             ++ generateBasicBullet (bullet_fac b + 292.5) (bullet_pos b)
             ++ generateBasicBullet (bullet_fac b + 337.5) (bullet_pos b)

startburst_step :: Bullet -> [Bullet]
startburst_step = slowToZero
                $ spinUpTo 12
                $ burst
                $ (\b -> [b])

tripleDamage :: (Bullet -> [Bullet]) -> (Bullet -> [Bullet])
tripleDamage f = \b ->
  [ b { bullet_step = f , damage = (damage b) * 3 } ]

accelerateTo_b :: Float -> (Bullet -> [Bullet]) -> (Bullet -> [Bullet])
accelerateTo_b speed f = \b ->
  if (bullet_vel b) == speed then [ b { bullet_step = f } ]
  else [ b { bullet_vel = min (bullet_vel b + 30) speed } ]

reverseDirection :: (Bullet -> [Bullet]) -> (Bullet -> [Bullet])
reverseDirection f = \b ->
  [ b { bullet_fac = bullet_fac b + 180 , bullet_dir = bullet_dir b + 180 , bullet_step = f } ]

boomerang_step :: Bullet -> [Bullet]
boomerang_step = slowToZero
               $ reverseDirection
               $ tripleDamage
               $ accelerateTo_b 1200
               $ (\b -> [b])
---------------------------------------------------------------------------------------
-- Enemy Building Blocks. Fun stuff here
---------------------------------------------------------------------------------------
accelerateTo_go :: Float -> (GameObject -> [GameObject]) -> (GameObject -> [GameObject])
accelerateTo_go speed f = \go ->
  if (velocity go) == speed then [ go { step = f } ]
  else [ go { velocity = min (velocity go + 30) speed } ]
-- Once again, this assumes the game runs perfectly at 60 fps. probably bad juice.
waitFor :: Float -> (GameObject -> [GameObject]) -> (GameObject -> [GameObject])
waitFor time f = \go ->
  if time == 0 then
    [go { step = f }]
  else 
    [go { step = waitFor (time - 1) f }]

seekerTurn :: (GameObject -> [GameObject]) -> (GameObject -> [GameObject])
seekerTurn f = \go ->
  [go { facing = facing go + 137.5 , step = f }]

seekerStop :: (GameObject -> [GameObject]) -> (GameObject -> [GameObject])
seekerStop f = \go ->
  [go { step = f , velocity = 0 }]

seeker_step :: GameObject -> [GameObject]
seeker_step = seekerTurn
            $ waitFor 120
            $ accelerateTo_go 1200
            $ waitFor 120
            $ seekerStop
            $ seeker_step

carrierContinue :: Float -> (GameObject -> [GameObject]) -> (GameObject -> [GameObject])
carrierContinue time f = \go ->
  if time == 0 then
    [go { step = f } ]
  else
    [go { step = carrierContinue (time - 1) f , facing = facing go + 1 }]

releaseZoomer :: (GameObject -> [GameObject]) -> (GameObject -> [GameObject])
releaseZoomer f = \go ->
  [go { step = f }, generateZoomer (position go) ((facing go) + 90)]

carrier_step :: GameObject -> [GameObject]
carrier_step = carrierContinue 300
             $ releaseZoomer
             $ carrier_step